massgov/mayflower
# React/DP-14424: TabContainer, Tab Refactor

## Description
TabContainer and Tab needed a refactor so that they could better handle server side rendering, and not track Tab content within TabContainer. To that end, [React Portals](https://reactjs.org/docs/portals.html) are being used. To support server side rendering, the [react-portal](https://github.com/tajo/react-portal) package was added. This is simply a portals helper and does not provide the actual portals API, as that's directly from React.

The TabBody component was created for rendering Tab's children directly into TabContainer's tab body div. This is accomplished using [MutationObserver,](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) allowing the TabBody component to watch for DOM changes to a specific DOM node. When TabContainer's tab body div is rendered and added to the document, TabBody's state is updated with that DOM node, which is then used to render Tab's children.

This PR should be used [to test a similar PR on the budget repo](https://github.com/massgov/budget/pull/306).
## Related Issue / Ticket

- [JIRA issue](https://jira.mass.gov/browse/DP-14424)

## Steps to Test
1. Pull the branch, then: `nvm use`, `npm install`, `npm run start`.
2. Browse to Organisms -> TabContainer and verify that none of the visible functionality has changed from before this branch was made.
